### PR TITLE
[Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51427 [Gradient Compression] Check start_PowerSGD_iter > 1
* **#51270 [Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations**
* #51400 [Resubmission][Gradient Compression] Refactor default_hooks.py and powerSGD_hook.py by creating a util function that make a vanilla allreduce future

Similar to #50973, allow the batched version to run vanilla allreduce for the first K iterations.

This may be useful if the batched version can be applied to some use cases where the accuracy requirement is not very strict.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26077709](https://our.internmc.facebook.com/intern/diff/D26077709/)

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 1]
```diff
@@ -423,6 +423,13 @@ def batched_powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
 
     # The input tensor is a flattened 1D tensor.
     input_tensor = bucket.get_tensors()[0]
+
+    # Run vanilla allreduce in the first `start_powerSGD_iter` iterations.
+    if state.iter < state.start_powerSGD_iter:
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:428]: Do we have unittests in place for `batched_powerSGD_hook`? If not, can we add them?

Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:428]: Batched powerSGD is already tested by `test_powerSGD_ddp_comm_hook_nccl` in test_c10d.py.

**Review by rohan-varma — COMMENTED** [review]
I think the diff below this one was reverted, and this one is also failing tests: https://app.circleci.com/pipelines/github/pytorch/pytorch/266089/workflows/a036f791-01c8-4538-90eb-a4e40234b8c3/jobs/10492151. Do you need to resubmit the previous diff first?

**Comment by SciPioneer** [comment]
> I think the diff below this one was reverted, and this one is also failing tests: https://app.circleci.com/pipelines/github/pytorch/pytorch/266089/workflows/a036f791-01c8-4538-90eb-a4e40234b8c3/jobs/10492151. Do you need to resubmit the previous diff first?

Created #51400 for the resubmission, and will submit that PR first.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 2]
```diff
@@ -399,6 +399,10 @@ def batched_powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
     7) Computes M, which is approximately equal to PQ^T.
     8) Truncates the input tensor to the original length.
 
+    Note that this communication hook enforces vanilla allreduce for the first `state.start_powerSGD_iter` iterations.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:402]: Do we have the default value for `state.start_powersgd_iter` documented in the docs for powerSGDState? Would be nice to ensure we have that, and maybe also specify the default start iteration here.

**Review by rohan-varma — APPROVED** [review]

**Pull request closed without merging**